src/persist/src/operators/stream.rs

```diff
@@ -41,46 +50,215 @@ where
+    /// Passes through each element of the stream and seals the given collection (the `write`
+    /// handle) when the input frontier advances.
+    ///
+    /// This does not wait for the seal to succeed before passing through the data. We do, however,
+    /// wait for the seal to be successful before allowing the frontier to advance. In other words,
+    /// this operator is holding on to capabilities as long as seals corresponding to their
+    /// timestamp are not done.
+    fn seal(
+        &self,
+        name: &str,
+        write: StreamWriteHandle<K, V>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    );
+}
+
+impl<G, K, V> Seal<G, K, V> for Stream<G, ((K, V), u64, isize)>
+where
+    G: Scope<Timestamp = u64>,
+    K: TimelyData + Codec,
+    V: TimelyData + Codec,
+{
+    fn seal(
+        &self,
+        name: &str,
+        write: StreamWriteHandle<K, V>,
     ) -> (
         Stream<G, ((K, V), u64, isize)>,
         Stream<G, (String, u64, isize)>,
     ) {
-        let (write, read) = token;
-
-        let (ok_new, err_new) = self.ok_err(
-            move |((k, v), ts, diff)| -> Result<((K, V), u64, isize), (String, u64, isize)> {
-                let res = write.write(&[((k.clone(), v.clone()), ts, diff)]);
-                // TODO: Don't do this synchronously.
-                res.recv().map_err(|err| {
-                    let err_str = format!("persisting data: {}", err);
-                    (err_str, ts, 1)
-                })?;
-                Ok(((k, v), ts, diff))
-            },
-        );
-
-        // Replay the previously persisted data, if any.
-        let (ok_previous, err_previous) = match read.snapshot() {
-            Err(err) => (
-                operator::empty(&self.scope()),
-                // TODO: Figure out how to make these retractable.
-                vec![(format!("replaying persisted data: {}", err), 0, 1)]
-                    .to_stream(&mut self.scope()),
-            ),
-            Ok(snapshot) => operators::replay(&mut self.scope(), snapshot),
-        };
-
-        let ok_all = ok_previous.concat(&ok_new);
-        let err_all = err_previous.concat(&err_new);
-        (ok_all, err_all)
+        let operator_name = format!("seal({})", name);
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut data_input = seal_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = seal_op.new_output();
+        let (mut error_output, error_output_stream) = seal_op.new_output();
+
+        let mut data_buffer = Vec::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+
+        let active_seal_operator = self.scope().index() == 0;
+
+        seal_op.build(move |mut capabilities| {
+            let mut err_cap = Some(capabilities.pop().expect("missing capability"));
```

Comment:
I actually don't follow what's happening here, probably because I haven't directly used OperatorBuilder. perhaps some context?

Reply:
We need to maintain a capability that allows us to send errors downstream. We can either do that by keeping and managing the initial capability or by retaining capabilities from input data.

I'll push a change that retains input capabilities, because that seems to be more idiomatic. I don't know if it simplifies things, though... :sweat_smile:

Reply:
It's probably best to to a quick huddle (in slack or elsewhere) about this.

Reply:
Oh I think you're overestimating the sophistication of my question :). After seeing your question in dataflow slack, I think the answer I was looking for here is "these are capabilities that correspond to the operator inputs".

Like frank said there, I think the right thing to do is keep whichever capabilities make the most semantic sense, in this case the operator is not really a source, so what you have here is correct I think

Reply:
Oh, but these capabilities are capabilities for the outputs. So I was doing the thing that Frank recommended we don't do for non-source operators. (to many double-no's in there...)

I pushed a change that retains input capabilities, but the problem with that is that we can only seal once we have received data at least once. Before that, we didn't yet have a chance to retain a capability.

Side note: the code I use now for handling the capabilities comes mostly from the dd arrange operator. It's somewhat complicated but very general.